File: programming-skills/unit-testing/unittest/Java/src/dna/Sequence.java
package dna;

import java.util.HashMap;

public class Sequence {
    /** Map from nucleotides to molecular weights. */
    public static final HashMap<Character, Double> WEIGHTS;
    /** Nucleotides in the sequence. */
    private String nucleotides;
    /** Initialise the WEIGHTS map. */
    static {
        WEIGHTS = new HashMap<Character, Double>();
        WEIGHTS.put('A', 131.2);
        WEIGHTS.put('C', 289.2);
        WEIGHTS.put('G', 329.2);
        WEIGHTS.put('T', 304.2);
    }
    
    /**
     * Create a new DNA sequence.
     * @param nucleotides Any lower-case characters are converted to
     * upper-case. 
     * @throws IllegalArgumentException if nucleotides contains a
     * character that is not A, C, G, T (upper or lower case).
     */
    public Sequence(String nucleotides) {
        if (! isValid(nucleotides)) {
            throw new IllegalArgumentException(nucleotides);
        }
        this.nucleotides = nucleotides.toUpperCase();
    }

    /**
     * Is a string a valid DNA sequence consisting of zero or more
     * characters A, C, G, T (upper or lower case)
     * @param nucleotides string to check.
     * @return true if it is valid, false otherwise.
     */
    public static boolean isValid(String nucleotides) {
        char[] chars = nucleotides.toUpperCase().toCharArray();
        boolean isValid = true;
        for (char c: chars) {
            isValid = isValid && WEIGHTS.containsKey(c);
        }
        return isValid;
    }
    
    /**
     * Get the DNA sequence.
     * @return DNA sequence. All nucleotides will be in upper-case.
     */
    public String getNucleotides() {
        return this.nucleotides;
    }

    /**
     * Get the molecular weight of this DNA sequence.
     * @return molecular weight formed by summing the weight of each
     * nucleotide in turn.
     */
    public double getWeight() {
        return calculateWeight(this);
    }
    
    /**
     * Calculate the molecular weight of a DNA sequence.
     * @param sequence DNA sequence.
     * @return molecular weight formed by summing the weight of each
     * nucleotide in turn.
     */
    public static double calculateWeight(Sequence sequence) {
        char[] chars = sequence.getNucleotides().toCharArray();
        double weight = 0;
        for (char c: chars) {
            weight += WEIGHTS.get(c);
        }
        return weight;
    }
    
    public static void main(String[] arguments) {
        Sequence sequence = new Sequence(arguments[0]);
        System.out.println(sequence.getWeight());
    }
}

File: programming-skills/unit-testing/unittest/Python/src/sequence.py
import sys

class Sequence:
    """
    DNA sequence consisting of A, C, G, T sequences.
    """

    WEIGHTS = {'A': 131.2, 'C': 289.2, 'G': 329.2, 'T': 304.2}
    """dict of str or unicode to float: nucleotide molecular weights"""

    def __init__(self, sequence=""):
        """Initialise Sequence with a string. The string must only
        contain letters in the set a,A,c,C,g,G,t,T.
        
        :param sequence: sequence
        :type sequence: str or unicode
        :raises AssertionError: if sequence contains an invalid letter
        """
        assert Sequence.is_valid(sequence), \
            "Sequence should only contain A, C, G and T"
        self.nucleotides = sequence.upper()

    @staticmethod
    def is_valid(nucleotides):
        """Is a given string a valid sequence of nucleotides? Does it
        only contain letters in the set a,A,c,C,g,G,t,T?
        
        :param nucleotides: nucleotides
        :type nucleotides: str or unicode
        :return: True if so, else False
        :rtype: bool
        """
        upper = nucleotides.upper()
        is_valid = True
        for c in upper:
            is_valid = is_valid and c in Sequence.WEIGHTS
        return is_valid

    @property
    def nucleotides(self):
        """Get nucleotides of this sequence, in upper-case.
    
        :return: nucleotides
        :rtype: str or unicode
        """
        return self.nucleotides

    def get_weight(self):
        """Calculate molecular weight of this DNA sequence.
    
        :return: molecular weight
        :rtype: float
        """
        return Sequence.calculate_weight(self)

    @staticmethod
    def calculate_weight(sequence):
        """Calculate molecular weight of a DNA sequence.
        
        :param sequence:
        :type sequence: Sequence
        :return: molecular weight
        :rtype: float
        """
        weight = 0
        for c in sequence.nucleotides:
            weight += Sequence.WEIGHTS[c]
        return weight

if __name__ == "__main__":
    sequence = Sequence(sys.argv[1])
    print(sequence.get_weight())
